core-schema/src/schema.rs

```diff
@@ -0,0 +1,585 @@
+    ///     .provide("https://spec.example.com/A", Version(1, 7), "impl for A 1.7");
+    /// let schema = Schema::parse(r#"schema @core(feature: "https://lib.apollo.dev/core/v0.1") @core(feature: "https://spec.example.com/A/v1.0") { query: Query }"#)?;
+    ///
+    /// let max = schema
+    ///   .activations(&impls)
+    ///   .filter_map(|(_request, mut impls)| impls.bounds())
+    ///   .collect::<Vec<_>>();
+    ///
+    /// assert_eq!(max, vec![(
+    ///     (&Version(1, 2), &"impl for A 1.2"), (&Version(1, 7), &"impl for A 1.7")
+    /// )]);
+    /// # Ok::<(), GraphQLParseError>(())
+    /// ```
+    pub fn activations<'impls, T>(
+        &'impls self,
+        implementations: &'impls Implementations<T>,
+    ) -> impl Iterator<
+        Item = (
+            &'a Feature,
+            Find<'impls, T, impl Iterator<Item = Found<'impls, T>>>,
+        ),
+    > + 'impls
+    where
+        'impls: 'a,
+        T: 'impls,
+    {
+        self.features
+            .iter()
+            .map(move |request| (request, implementations.find_feature(request)))
+    }
+}
+
+/// Drain all items from a `Vec<T>` which match `pred` and collect the results
+///
+/// Equivalent to (and probably slower than) `vec.drain_filter(pred).collect()`,
+/// which we should switch to once it stabilizes.
+fn drain_filter_collect<T, F: Fn(&T) -> bool>(vec: &mut Vec<T>, pred: F) -> Vec<T> {
+    let mut i = 0;
+    let mut collected = vec![];
+    while i != vec.len() {
+        if pred(&vec[i]) {
+            collected.push(vec.remove(i));
+        } else {
+            i += 1;
+        }
+    }
+    collected
+}
+
+// Given a `SchemaDefinition`, locate the first `@core` request, which must be for
+// the core spec itself.
+fn bootstrap(schema: &SchemaDefinition) -> Option<Feature> {
+    schema
+        .directives
+        .iter()
+        // Scan requests which parsed without error
+        .filter_map(|dir| Feature::from_directive(dir).map(|res| (dir, res)))
+        .filter_map(|(dir, res)| res.ok().map(|req| (dir, req)))
+        .find_map(|(dir, req)| {
+            if
+            // We're looking for bootstrap @core directives, so select only
+            // those requests whose requested name matches the directive name...
+            dir.name == req.name &&
+                // ...and which are requesting a the core spec...
+                req.spec.identity == CORE.identity &&
+                // ...at a supported version.
+                CORE.version.satisfies(&req.spec.version)
+            {
+                Some(req)
+            } else {
+                None
+            }
+        })
+}
+/// Validation errors which may occur on a `Schema`
+#[derive(Debug, Error, PartialEq, Eq)]
+pub enum SchemaError {
+    /// Machine schemas must contain exactly one schema definition.
+    #[error("no schema definitions found in document")]
+    NoSchemas,
+
+    /// No reference to the core spec found
+    #[error("no reference to the core spec found")]
```

Comment:
technically this is more like no reference to an understood version of the core spec is found. maybe the error can actually show the identity and version (or just the URL) that is understood?